### PR TITLE
Proposal: F3 - Filecoin storage deals reliability and UX

### PR DESCRIPTION
**Pull request by raulk**

### CODE REVIEW

File: proposals/0012-f3-deals-reliability-ux.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+2. display miners (miner directory in the future).
+3. select several miners.
+4. get a quotes from those miners.
+5. select a miner.
+6. initiate a deal.
+7. etc.
+
+However, if we consider the end-to-end flow to span all the way through to on-chain deal confirmation and sealing, interactivity might be suboptimal, as the end-to-end flow is in the order of hours.
+
+Alternatively, the CLI could produce a ticket number, and through a different command, the user could enquire about progress on that ticket number.
```
Comment (dirkmc) [inline proposals/0012-f3-deals-reliability-ux.md:61]: > the CLI could produce a ticket number

At the moment the CLI lets you track deal progress using a deal ID. It's a little unwieldy as you need to use several commands to get all the information about the deal, eg `lotus client list-deals` cross referenced with `lotus client list-transfers` using the `piece CID` as the field to join the two outputs

File: proposals/0012-f3-deals-reliability-ux.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+> We decided to name this Project F3 in a wink to shooting to the same level of usability of Amazon S3 (albeit with notable differences).
+
+#### Assumptions & hypotheses
+
+_What must be true for this project to matter?_
+
+1. Users struggle to grasp the concepts and components that comprise the deal-making process end to end.
+2. Users don't have intuition of on what happens under the hood (not the details, but the major phases).
+3. New users desist, or end up frustrated.
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:37]: desist what ?

I think you mean "cease experimenting with Filecoin" ?

File: proposals/0012-f3-deals-reliability-ux.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+4. get a quotes from those miners.
+5. select a miner.
+6. initiate a deal.
+7. etc.
+
+However, if we consider the end-to-end flow to span all the way through to on-chain deal confirmation and sealing, interactivity might be suboptimal, as the end-to-end flow is in the order of hours.
+
+Alternatively, the CLI could produce a ticket number, and through a different command, the user could enquire about progress on that ticket number.
+
+Taking a step back, user-friendly CLI commands are nice, but if we want to push the UX even further, it would be  desirable to have a companion graphical web-based deals dashboard that one can point to a node to display the status and details of all deals. An interface like could be enriched with instructional cues to educate the user progressively.
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:63]:
```suggestion
Taking a step back, user-friendly CLI commands are nice, but if we want to push the UX even further, it would be  desirable to have a companion graphical web-based deals dashboard that one can point to a node to display the status and details of all deals. An interface like **what ?** could be enriched with instructional cues to educate the user progressively.
```

File: proposals/0012-f3-deals-reliability-ux.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+The Spark time could help prototype this web UI ;-)
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:65]:
```suggestion
The Spark team could help prototype this web UI ;-)
```

File: proposals/0012-f3-deals-reliability-ux.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+This problem statement addresses massive painpoints in the primary use case that Filecoin was built for (file storage and retrieval). Not executing on this translates into first-time user failure and, consequently, bounce.
+
+#### Leverage
+
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+🎯🎯🎯
+
+High. For seeveral reasons: this project will likely land a bunch of refactors that will make the affected areas of the codebase (markets, data transfer, graphsync) more accessible to new developers. Additionally, having a consistent high deal success rate, with a seamless and great UX that gets good user feedback, will get the team much satisfaction, which will translate into higher drive to continue driving improvements. Finally, we hope that new users/devs arriving to Filecoin and having success when making their first deal will translate into interest and hunger for more, and ultimately into some form of contribution.
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:81]: This paragraph is poetry :)

File: proposals/0012-f3-deals-reliability-ux.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+10. We have very solid evidence that the deals flow is... suboptimal.
+
+## Project definition
+
+#### Brief plan of attack
+
+**UX track**
+
+1. Study the user flow in detail. Inventorise the specific elements that are confusing users.
+2. Initiate discussions about user-friendlier terminology.
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:96]:
```suggestion
2. Initiate discussions about user-friendly terminology.
```

File: proposals/0012-f3-deals-reliability-ux.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+5. One-stop set of JSON-RPC operations that encapsulate the new behaviour with no frills (without leaking implementation details as much as possible).
+
+**Technical reliability track**
+
+0. Do a full inventory of outstanding issues; get to the bottom of the long-standing mystical issues; analyse; fix. Some known tasks:
+1. Testing: Design and implement test suites to reproduce the concrete issues and serve as regression tests. Nightly tests for deals.
+2. Fix memory footprint issues.
+3. Fix resumability of data transfers.
+4. Consider p2p file-oriented transfer protocol to cut out complexity for now.
+5. Identify areas of the deals flow code that need more informative logging (eg data transfer restarts / connection retries)
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:108]:
```suggestion
5. Identify areas of the deals flow code that need more informative logging (eg data transfer restarts / connection retries).
```

File: proposals/0012-f3-deals-reliability-ux.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+2. Initiate discussions about user-friendlier terminology.
+3. Design a new CLI UX, prototype it, iterate.
+4. Implement CLI UX changes (and underlying code changes)
+5. One-stop set of JSON-RPC operations that encapsulate the new behaviour with no frills (without leaking implementation details as much as possible).
+
+**Technical reliability track**
+
+0. Do a full inventory of outstanding issues; get to the bottom of the long-standing mystical issues; analyse; fix. Some known tasks:
+1. Testing: Design and implement test suites to reproduce the concrete issues and serve as regression tests. Nightly tests for deals.
+2. Fix memory footprint issues.
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:105]: how ? maybe link to RFCs, issues or proposals ?

File: proposals/0012-f3-deals-reliability-ux.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+3. Fix resumability of data transfers.
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:106]: link RFCs/proposals.

File: proposals/0012-f3-deals-reliability-ux.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+4. Consider p2p file-oriented transfer protocol to cut out complexity for now.
+5. Identify areas of the deals flow code that need more informative logging (eg data transfer restarts / connection retries)
+6. Add logs and journal events.
+7. Fix miner datastore bloat.
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:110]: link proposals.

File: proposals/0012-f3-deals-reliability-ux.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+#### What does done look like?
```
Comment (aarshkshah1992) [inline proposals/0012-f3-deals-reliability-ux.md:112]: The done & success sections don't mention anything about visibility into completion of datastore bloat fix, memory footprint issues & resum-ability of data transfers..

How will we measure that we have achieved those ?

**Review by aarshkshah1992 — COMMENTED** [review]
@raulk @dirkmc @nonsense 

1. Link concrete proposals/POCs/RFCs/previous work when we talk of fixing data store bloat, reducing memory footprint and resuming deal transfer. Statements such as "reduce memory footprint" without a link to a proposed solution sound too abstract. Do we have a plan for this ?
2. Enlist how we will measure done/success for all the above mentioned fixes i.e. compare go profiles, test specific data transfer scenarios that are failing, compare data-store sizes after doing a fixed set of deals etc.

Without mentioning any previously thought out approaches to these problems, it's hard to arrive at an estimate of how long the project will take because we will need time to design the solutions too.

File: proposals/0012-f3-deals-reliability-ux.md (hunk 3 above)
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:63]:
```suggestion
Taking a step back, user-friendly CLI commands are nice, but if we want to push the UX even further, it would be desirable to have a companion graphical web-based deals dashboard that one can point to a node to display the status and details of all deals. An interface like this could be enriched with instructional cues to educate the user progressively.
```

File: proposals/0012-f3-deals-reliability-ux.md (hunk 2 above)
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:37]: Yep, synonym of "give up".

File: proposals/0012-f3-deals-reliability-ux.md (hunk 8 above)
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:105]: Too early for the technical definition -- that said, there's a lot, but it's being incubated in Google Docs. Once we're ready to publish, we will, and hopefully someone will remember to come back and crosslink here.

File: proposals/0012-f3-deals-reliability-ux.md (hunk 9 above)
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:106]: See above.

File: proposals/0012-f3-deals-reliability-ux.md (hunk 10 above)
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:110]: See above.

**Comment by nonsense** [comment]
@aarshkshah1992 this document is about stating the problems, and not the exact plan on how we will tackle them.

> Link concrete proposals/POCs/RFCs/previous work when we talk of fixing data store bloat, reducing memory footprint and resuming deal transfer. Statements such as "reduce memory footprint" without a link to a proposed solution sound too abstract. Do we have a plan for this ?

There aren't any POCs or RFCs as far as I am aware, but your Filecoin node needing 30GB of RAM to send a 30GB file to a given miner is a problem, and what this item refers to.

> Enlist how we will measure done/success for all the above mentioned fixes i.e. compare go profiles, test specific data transfer scenarios that are failing, compare data-store sizes after doing a fixed set of deals etc.

This project also includes an entry to write tests for these improvements. So far we've discovered those problem by running nodes and miners, and Filecoin community members are also complaining about these issues.

> Without mentioning any previously thought out approaches to these problems, it's hard to arrive at an estimate of how long the project will take because we will need time to design the solutions too.

I am under the impression that project proposals are about stating urging problems that we feel we can contribute to, and not way out a full task plan on how we will do that. Doing this is probably 80-90% of the work and will happen during the project.

File: proposals/0012-f3-deals-reliability-ux.md [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+1. Study the user flow in detail. Inventorise the specific elements that are confusing users.
+2. Initiate discussions about user-friendlier terminology.
+3. Design a new CLI UX, prototype it, iterate.
+4. Implement CLI UX changes (and underlying code changes)
+5. One-stop set of JSON-RPC operations that encapsulate the new behaviour with no frills (without leaking implementation details as much as possible).
+
+**Technical reliability track**
+
+0. Do a full inventory of outstanding issues; get to the bottom of the long-standing mystical issues; analyse; fix. Some known tasks:
+1. Testing: Design and implement test suites to reproduce the concrete issues and serve as regression tests. Nightly tests for deals.
```
Comment (momack2) [inline proposals/0012-f3-deals-reliability-ux.md:104]: Sounds like @alanshaw was also proposing carving this out

File: proposals/0012-f3-deals-reliability-ux.md [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+2. Fix memory footprint issues.
+3. Fix resumability of data transfers.
+4. Consider p2p file-oriented transfer protocol to cut out complexity for now.
```
Comment (momack2) [inline proposals/0012-f3-deals-reliability-ux.md:107]: Scope is starting to get potentially very large here...

Comment (momack2) [reply proposals/0012-f3-deals-reliability-ux.md:107]: Can you break up this track into 2 separate buckets for must do vs potential extensions?

File: proposals/0012-f3-deals-reliability-ux.md [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+1. A new version of Lotus (and associated modules) that achieves the success criteria.
+2. A new interactive CLI command (`lotus store`) that guides the user through the process of performing a storage deal, and provides visibility and a polished flow.
+3. A simplified, porcelain S3-like JSON-RPC API that covers the deals flow and provides the logical sequence of steps outlined in previous sections, under a refined, well-documented, and opaque API that doesn't leak Filecoin implementation details.
+4. (Optionally) A prototype for a deals dashboard web UI (in collaboration with The Spark).
+
+#### What does success look like?
+
+_Success means impact. How will we know we did the right thing?_
+
+- 80% of deals initiated via the CLI result in success, as long as prerequisites are met (enough balance, synced node, etc.) under a variety of conditions (e.g. connection losses, slow bandwidth, concurrent deals).
```
Comment (momack2) [inline proposals/0012-f3-deals-reliability-ux.md:125]: Maybe: # of deals made on the network, # of clients who have made at least 1 successful deal, number of projects actively storing data on Filecoin

File: proposals/0012-f3-deals-reliability-ux.md [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+- Most open issues related to deals on the Filecoin issue tracker are closed, and not reopened, and new ones related to deals are not created shortly thereafter (i.e. we don't introduce new problems).
+- Users give consistent positive feedback about the usability and visibility characteristics of deal-making.
+
+#### Counterpoints & pre-mortem
+
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+N/A. There's pretty good consensus that this is a high impact project.
```
Comment (momack2) [inline proposals/0012-f3-deals-reliability-ux.md:133]: I think the risk here is scope creep. Motivation is great but it can become a “boil the ocean” project with a million sub-project components. How do we scope this down to be maximally valuable and efficient?

File: proposals/0012-f3-deals-reliability-ux.md [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+
+#### Future opportunities
+
+We won't be focusing on the deals dashboard web UI right off the bat, but we could work with The Spark to deliver a prototype, and then open up a bounty/grant for full development.
+
+## Required resources
+
+#### Effort estimate
+
+Large/XLarge, depending on final scope and solution design.
```
Comment (momack2) [inline proposals/0012-f3-deals-reliability-ux.md:153]: This is my concern. Can we boil down to just “large” by getting really clear on what’s in scope? Concerned this secretly has 3-4 separate projects buried in it

**Review by momack2 — COMMENTED** [review]
Love the motivation, but I think the scope needs more clarity (Ex, please scope large “foundations” workstreams to be their own projects vs bundled in here

File: proposals/0012-f3-deals-reliability-ux.md (hunk 14 above)
Comment (BigLep) [reply proposals/0012-f3-deals-reliability-ux.md:125]: Do we have metrics today on deals initiated via the CLI and then which portion of those are successful?  If not, should we add that to the project plan to first implement/release those metrics so we can see the impact of the F3 work?

File: proposals/0012-f3-deals-reliability-ux.md [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+
+`lotus store <path>`
+
+1. add file to the repo.
+2. display miners (miner directory in the future).
+3. select several miners.
+4. get a quotes from those miners.
+5. select a miner.
+6. initiate a deal.
+7. etc.
```
Comment (BigLep) [inline proposals/0012-f3-deals-reliability-ux.md:57]: This sounds like a workflow.  A couple of questions:
1. What are the UX implications if the lotus process dies part way through the workflow state?
2. Will the the workflow state persisted anywhere?  Per comment below from dirkmc@ it sounds like the deal state is stored somewhere.   I'm asking about this from how to potentially support "dashboard that one can point to a node to display the status and details of all deals" in the future.
(feel free to disregard if these comments/questions aren't relevant)

File: proposals/0012-f3-deals-reliability-ux.md [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+
+However, if we consider the end-to-end flow to span all the way through to on-chain deal confirmation and sealing, interactivity might be suboptimal, as the end-to-end flow is in the order of hours.
+
+Alternatively, the CLI could produce a ticket number, and through a different command, the user could enquire about progress on that ticket number.
+
+Taking a step back, user-friendly CLI commands are nice, but if we want to push the UX even further, it would be desirable to have a companion graphical web-based deals dashboard that one can point to a node to display the status and details of all deals. An interface like this could be enriched with instructional cues to educate the user progressively.
```
[the remark this answers is not in the document]
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:63]: Yes! On my plate to submit a project pitch. cc @alanshaw And we shall partner closely ;-)

File: proposals/0012-f3-deals-reliability-ux.md (hunk 12 above)
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:104]: We intended this to be local, scoped integration tests that reproduce the memory, datastore, etc. issues -- not distributed, multipartipant, higher-level, acceptance end-to-end tests, which will necesarily exercise many code paths in Lotus.

File: proposals/0012-f3-deals-reliability-ux.md (hunk 13 above)
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:107]: Yep, we will follow up with an execution plan that divides this project into incremental milestones.

File: proposals/0012-f3-deals-reliability-ux.md (hunk 16 above)
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:153]: It does, but I think we're better off creating a phased milestone approach to an XLarge project, than breaking it up into interdependent smaller projects just to represent progression. Some things can be broken out:

* Deals web dashboard (I'll submit a pitch for this).
* Deals end-to-end testing (#28).

File: proposals/0012-f3-deals-reliability-ux.md (hunk 14 above)
Comment (raulk) [reply proposals/0012-f3-deals-reliability-ux.md:125]: @momack2 how would you feel about dusting off the dealbot we used in Space Race, and "letting it loose" on mainnet?

**Comment by mikeal** [comment]
There’s a number of unaddressed review comments but I think we’ve actually moved forward with some if this proposal already so I’m not sure how to proceed with the PR
